interfaces/ldconfig/backend.go

```diff
@@ -0,0 +1,121 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ldconfig
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/timings"
+)
+
+// Backend is responsible for maintaining ldconfig cache.
+type Backend struct{}
+
+var _ = interfaces.SecurityBackend(&Backend{})
+
+// Initialize does nothing for this backend.
+func (b *Backend) Initialize(opts *interfaces.SecurityBackendOptions) error {
+	return nil
+}
+
+// Name returns the name of the backend.
+func (b *Backend) Name() interfaces.SecuritySystem {
+	return "ldconfig"
+}
+
+// Setup will make the ldconfig backend generate the needed
+// configuration files and re-create the ld cache.
+//
+// If the method fails it should be re-tried (with a sensible strategy) by the caller.
+func (b *Backend) Setup(appSet *interfaces.SnapAppSet, opts interfaces.ConfinementOptions, repo *interfaces.Repository, tm timings.Measurer) error {
+	snapName := appSet.InstanceName()
+	// Get the snippets that apply to this snap
+	spec, err := repo.SnapSpecification(b.Name(), appSet, opts)
+	if err != nil {
+		return fmt.Errorf("cannot obtain ldconfig specification for snap %q: %s",
+			snapName, err)
+	}
+
+	return b.setupLdconfigCache(spec.(*Specification))
+}
+
+// Remove removes modules ldconfig files specific to a given snap.
+// This method should be called after removing a snap.
+//
+// If the method fails it should be re-tried (with a sensible strategy) by the caller.
+func (b *Backend) Remove(snapName string) error {
+	// TODO this will not be called for the rootfs case where the system
+	// has an implicit plug, but needs to be revisited for snaps.
+	return nil
+}
+
+// NewSpecification returns a new specification associated with this backend.
+func (b *Backend) NewSpecification(*interfaces.SnapAppSet,
+	interfaces.ConfinementOptions) interfaces.Specification {
+	return &Specification{}
+}
+
+// SandboxFeatures returns the list of features supported by snapd for ldconfig.
+func (b *Backend) SandboxFeatures() []string {
+	return []string{"mediated-ldconfig"}
+}
+
+func (b *Backend) setupLdconfigCache(spec *Specification) error {
+	ldConfigDir := dirs.SnapLdconfigDir
+	if err := os.MkdirAll(ldConfigDir, 0755); err != nil {
+		return fmt.Errorf("cannot create directory for ldconfig files %q: %s", ldConfigDir, err)
+	}
+
+	// TODO this considers only the case when the libraries are exposed to
+	// the rootfs. For snaps, we will create files in
+	// /var/lib/snapd/ldconfig/ that will be used to generate a cache
+	// specific to each snap.
+
+	allContent := map[string]osutil.FileState{}
+	for key, dirs := range spec.libDirs {
+		content := "# This file is automatically generated by snapd\n"
+		content += strings.Join(dirs, "\n")
+		content += "\n"
+
+		// File name is snap.<snap_name>.<slot_name>.conf
+		ldconfigPath := fmt.Sprintf("snap.%s.conf", key)
```

Comment:
based on what I found we want this to be based on the plug snap name I think. OTOH if the plug snap is snapd we could use a special name like snapd.conf or snap.system.conf ?

Reply:
Changed as discussed, this is ready for review now